examples/UNet/massachussets_road_segm.py

```diff
@@ -0,0 +1,340 @@
+        np.save("train_data.npy", data_train)
+        np.save("train_target.npy", target_train)
+        np.save("valid_data.npy", data_valid)
+        np.save("valid_target.npy", target_valid)
+        np.save("test_data.npy", data_test)
+        np.save("test_target.npy", target_test)
+    except:
+        print "something went wrong, maybe the download?"
+
+
+def plot_some_results(pred_fn, test_generator, BATCH_SIZE, PATCH_SIZE = 192):
+    fig_ctr = 0
+    ctr = 0
+    for data, seg in test_generator:
+        pred = pred_fn(data)
+        res = pred.argmax(-1).reshape(BATCH_SIZE, 1, PATCH_SIZE, PATCH_SIZE)
+        prob = pred.transpose(1, 0).reshape(2, BATCH_SIZE, PATCH_SIZE, PATCH_SIZE)[1]
+        for d, s, r, p in zip(data, seg, res, prob):
+            plt.figure(figsize=(12, 6))
+            plt.subplot(1, 4, 1)
+            plt.imshow(d.transpose(1,2,0))
+            plt.subplot(1, 4, 2)
+            plt.imshow(s[0])
+            plt.subplot(1, 4, 3)
+            plt.imshow(r[0])
+            plt.subplot(1, 4, 4)
+            plt.imshow(p)
+            plt.savefig("road_segmentation_result_%03.0f.png"%fig_ctr)
+            plt.close()
+            fig_ctr += 1
+        ctr += 1
+        if ctr > 5:
+            break
+
+
+def main():
+    # only download dataset once. This takes a while.
+    if not os.path.isfile("test_target.npy"):
+        # heuristic that I included to make sure the dataset is only donwloaded and prepared once
+        prepare_dataset()
+
+    # set some hyper parameters. You should not have to touch anything if you have 4GB or more VRAM
+    BATCH_SIZE = 24
+    N_EPOCHS = 30
+    N_BATCHES_PER_EPOCH = 100
+    N_BATCHES_PER_EPOCH_valid = 15
+    PATCH_SIZE = 128+64
+
+    # load the prepared data. They have been converted to np arrays because they are much faster to load than single image files.
+    # you will need some ram in order to have everything in memory.
+    # TODO is it worth the effort to refactor everything so that we use memmaps?
+    data_train = np.load("train_data.npy")
+    target_train = np.load("train_target.npy")
+    data_valid = np.load("valid_data.npy")
+    target_valid= np.load("valid_target.npy")
+    data_test = np.load("test_data.npy")
+    target_test = np.load("test_target.npy")
+
+    # we are using pad='same' for simplicity (otherwise we would have to crop our ground truth).
+    # Did not test for other paddings TODO
+    net = build_UNet(n_input_channels=3, BATCH_SIZE=BATCH_SIZE, num_output_classes=2, pad='same',
+                     nonlinearity=lasagne.nonlinearities.elu, input_dim=(PATCH_SIZE, PATCH_SIZE),
+                     base_n_filters=16, do_dropout=False)
+    output_layer = net["output"]
+
+    # if you wish to load pretrained weights you can uncomment this code and modify the file name
+    # if you want, use my pretained weights (got around 96% accuracy and a loss of 0.11 using excessive data
+    # augmentation)
```

Comment:
Can you elaborate? "using excessive data augmentation: rotation, cropping, flipping, x, y, z)"